### PR TITLE
Change about ln,in,softmax and creat hook about Transformer

### PR DESCRIPTION
**Pull request by timmywanttolearn**
adding hooks and funcitons about softmax, layernorm,prelu,instancenorm.
create hook and function about nn.Transformer
edit the softmax(which was wrong)

### CODE REVIEW

**Comment by timmywanttolearn** [comment]
我直接用中文说了，英文估计我说不明白。
1. 错误修正
首先我之前理解错误了关于Multiattention的问题，就是首先他要把embbeding的512(default)，分成num_head分，再去做q，k，v的运算。之前没有分开，导致Multiattention统计出错。现在的multition统计（通过print输出）是1082400.0 macs
第二个问题实在fft层(fed forward transform),我之前错误的理解成两个全链接层，其实是两个1维卷积层。
改过之后用提供的example，得出的对于一个512的向量结果为
44671552.0 44151808.0 
在encoderlayer里，计算量为3180576 macs
2.检验结果
改正这两个问题之后，我用tf.keras统计结果对比，代码如下.
```python
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
from tensorflow.python.framework.convert_to_constants import  convert_variables_to_constants_v2_as_graph
class TransformerBlock(layers.Layer):
    def __init__(self, embed_dim, num_heads, ff_dim, rate=0.1):
        super(TransformerBlock, self).__init__()
        self.att = layers.MultiHeadAttention(num_heads=num_heads, key_dim=embed_dim)
        self.ffn = keras.Sequential(
            [layers.Dense(ff_dim, activation="relu"), layers.Dense(embed_dim),]
        )
        self.layernorm1 = layers.LayerNormalization(epsilon=1e-6)
        self.layernorm2 = layers.LayerNormalization(epsilon=1e-6)
        self.dropout1 = layers.Dropout(rate)
        self.dropout2 = layers.Dropout(rate)

    def call(self, inputs, training):
        attn_output = self.att(inputs, inputs)
        attn_output = self.dropout1(attn_output, training=training)
        out1 = self.layernorm1(inputs + attn_output)
        ffn_output = self.ffn(out1)
        ffn_output = self.dropout2(ffn_output, training=training)
        return self.layernorm2(out1 + ffn_output)
class TokenAndPositionEmbedding(layers.Layer):
    def __init__(self, maxlen, vocab_size, embed_dim):
        super(TokenAndPositionEmbedding, self).__init__()
        self.token_emb = layers.Embedding(input_dim=vocab_size, output_dim=embed_dim)
        self.pos_emb = layers.Embedding(input_dim=maxlen, output_dim=embed_dim)

    def call(self, x):
        maxlen = tf.shape(x)[-1]
        positions = tf.range(start=0, limit=maxlen, delta=1)
        positions = self.pos_emb(positions)
        x = self.token_emb(x)
        return x + positions
embed_dim = 512  # Embedding size for each token
num_heads = 8  # Number of attention heads
ff_dim = 2048 # Hidden layer size in feed forward network inside transformer
maxlen = 1 
vocab_size = 1
inputs = layers.Input(shape=(maxlen,))
embedding_layer = TokenAndPositionEmbedding(maxlen, vocab_size, embed_dim)
x = embedding_layer(inputs)
transformer_block = TransformerBlock(embed_dim, num_heads, ff_dim)
x = transformer_block(x)
x = layers.GlobalAveragePooling1D()(x)
x = layers.Dropout(0.1)(x)
x = layers.Dense(20, activation="relu")(x)
x = layers.Dropout(0.1)(x)
outputs = layers.Dense(2, activation="softmax")(x)

model = keras.Model(inputs=inputs, outputs=outputs)
def get_flops(model):
    concrete = tf.function(lambda inputs: model(inputs))
    concrete_func = concrete.get_concrete_function(
        [tf.TensorSpec([1, *inputs.shape[1:]]) for inputs in model.inputs])
    frozen_func, graph_def = convert_variables_to_constants_v2_as_graph(concrete_func)
    with tf.Graph().as_default() as graph:
        tf.graph_util.import_graph_def(graph_def, name='')
        run_meta = tf.compat.v1.RunMetadata()
        opts = tf.compat.v1.profiler.ProfileOptionBuilder.float_operation()
        flops = tf.compat.v1.profiler.profile(graph=graph, run_meta=run_meta, cmd="op", options=opts)
        return flops.total_float_ops


print("The FLOPs is:{}".format(get_flops(model)) ,flush=True )
```
单个的encoder layer。
得出的关于一个512的向量的结果是4232000，大概是我的counter（结果的两倍-2000000），最为贴切的结果应该是两倍，因为macs和flops大概是两倍的关系。于是我拆解encoderlayer，encoderlayer主要有两个层，一个是fft，一个是Multiattention。
2.1
关于fft
改上述代码，发现keras统计flops为4228000，可以看到在keras中encoderlayer主要计算量分布在fft层。同时用print打出我的计数器fft层的计数2100000左右，大概是两倍的关系。我修改了输入的sequence，以及fft的forward_dimention。发现两者差距不大。因此我认为误差主要存在在Multiattention层。
2.2
```python
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
from tensorflow.python.framework.convert_to_constants import  convert_variables_to_constants_v2_as_graph
class TransformerBlock(layers.Layer):
    def __init__(self, embed_dim, num_heads, ff_dim, rate=0.1):
        super(TransformerBlock, self).__init__()
        self.att = layers.MultiHeadAttention(num_heads=num_heads, key_dim=embed_dim)

    def call(self, inputs, training):
        attn_output = self.att(inputs, inputs)
        return attn_output
class TokenAndPositionEmbedding(layers.Layer):
    def __init__(self, maxlen, vocab_size, embed_dim):
        super(TokenAndPositionEmbedding, self).__init__()
        self.token_emb = layers.Embedding(input_dim=vocab_size, output_dim=embed_dim)
        self.pos_emb = layers.Embedding(input_dim=maxlen, output_dim=embed_dim)

    def call(self, x):
        maxlen = tf.shape(x)[-1]
        positions = tf.range(start=0, limit=maxlen, delta=1)
        positions = self.pos_emb(positions)
        x = self.token_emb(x)
        return x + positions
embed_dim = 512  # Embedding size for each token
num_heads = 8  # Number of attention heads
ff_dim = 2048  # Hidden layer size in feed forward network inside transformer
maxlen = 1 
vocab_size = 1
inputs = layers.Input(shape=(maxlen,))
embedding_layer = TokenAndPositionEmbedding(maxlen, vocab_size, embed_dim)
x = embedding_layer(inputs)
transformer_block = TransformerBlock(embed_dim, num_heads, ff_dim)
x = transformer_block(x)

outputs = x

model = keras.Model(inputs=inputs, outputs=outputs)
def get_flops(model):
    concrete = tf.function(lambda inputs: model(inputs))
    concrete_func = concrete.get_concrete_function(
        [tf.TensorSpec([1, *inputs.shape[1:]]) for inputs in model.inputs])
    frozen_func, graph_def = convert_variables_to_constants_v2_as_graph(concrete_func)
    with tf.Graph().as_default() as graph:
        tf.graph_util.import_graph_def(graph_def, name='')
        run_meta = tf.compat.v1.RunMetadata()
        opts = tf.compat.v1.profiler.ProfileOptionBuilder.float_operation()
        flops = tf.compat.v1.profiler.profile(graph=graph, run_meta=run_meta, cmd="op", options=opts)
        return flops.total_float_ops


print("The FLOPs is:{}".format(get_flops(model)) ,flush=True )
```
发现对于单独的Multihead attention，keras结果是4193 flops。我认为这个是错误的，相比之下我的结果是1082400 macs。原因是在Multiattention层，pytorch使用了三个全链接层去得到qkv，这三者的计算量与embed_dim^2成正比，一定是远远大于4193的，更不用说qkv的计算与最后的全链接层。
因此我认为我的模型统计的多头注意力是正确的。
3.总结
最后，我使用了多组数据输入，发现sequence较大的时候Multition计算量反而比fft大得多，这是符合论文预测的。我曾经尝试用onnx模型从pytorch转换到tf，但是tf2.0的计数目前支持keras，无论是onnx还是pb，都无法把transformer转换为keras。这条路走不通。
